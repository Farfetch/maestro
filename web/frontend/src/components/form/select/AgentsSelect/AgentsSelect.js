import { Select } from "antd";
import { orderBy } from "lodash";
import PropTypes from "prop-types";

import { agentStatus } from "../../../../lib/api/models";
import AgentStatusBadge from "../../../badge/AgentStatusBadge";

const { Option } = Select;

const AgentsSelect = ({ value, agents, mode, onChange }) => (
  <Select
    mode={mode}
    allowClear
    showSearch
    filterOption={(input, option) => {
      const optionText = option.children.props.text.toLowerCase();
      const inputValue = input.toLowerCase();
      return optionText.includes(inputValue);
    }}
    optionFilterProp="children.props.text"
    style={{ width: "100%" }}
    placeholder="Select Agent from the list"
    onChange={onChange}
    value={value}
  >
    {orderBy(agents, "hostname", "asc").map((agent) => (
      <Option key={agent.id} value={agent.id}>
        <AgentStatusBadge
          agentStatus={agent.agentStatus}
          text={agent.hostname}
        />
      </Option>
    ))}
  </Select>
);

AgentsSelect.defaultProps = {
  onChange: () => {},
  mode: "single",
  value: ""
};

AgentsSelect.propTypes = {
  mode: PropTypes.oneOf(["single", "multiple"]).isRequired,
  agents: PropTypes.arrayOf(
    PropTypes.shape({
      agentStatus: PropTypes.oneOf(Object.values(agentStatus)),
      hostname: PropTypes.string.isRequired,
      id: PropTypes.string.isRequired
    })
  ).isRequired,
  onChange: PropTypes.func.isRequired
};

export default AgentsSelect;
