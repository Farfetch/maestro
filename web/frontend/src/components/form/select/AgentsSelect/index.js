import AgentsSelect from "./AgentsSelect";

export default AgentsSelect;
