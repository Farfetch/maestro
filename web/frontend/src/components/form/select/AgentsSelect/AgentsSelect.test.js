import { render } from "@testing-library/react";
import moment from "moment";
import React from "react";

import AgentsSelect from "./AgentsSelect";

describe("components/form/select/AgentsSelect", () => {
  const agents = [
    {
      id: "agent-id",
      hostname: "agent-hostname",
      ip: "test-ip",
      agentStatus: "AVAILABLE",
      createdAt: moment("2018-05-24T13:48:04.313000"),
      updatedAt: moment("2018-05-24T13:48:04.313000")
    },
    {
      id: "agent-id2",
      hostname: "agent-hostname2",
      ip: "test-ip2",
      agentStatus: "PROCESSING_EVENT",
      createdAt: moment("2018-05-24T13:48:04.313000"),
      updatedAt: moment("2018-05-24T13:48:04.313000")
    }
  ];
  ["single", "multiple"].map((mode) =>
    test(`should render AgentsSelect component with mode=${mode}`, async () => {
      const rendered = render(<AgentsSelect agents={agents} mode={mode} />);

      const badgeComponent = rendered.container;
      expect(badgeComponent.outerHTML).toMatchSnapshot();
    })
  );
});
