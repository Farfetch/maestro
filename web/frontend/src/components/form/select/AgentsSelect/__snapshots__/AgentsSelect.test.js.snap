// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/form/select/AgentsSelect should render AgentsSelect component with mode=multiple 1`] = `"<div><div class=\\"ant-select ant-select-multiple ant-select-allow-clear ant-select-show-search\\" style=\\"width: 100%;\\"><div class=\\"ant-select-selector\\"><div class=\\"ant-select-selection-overflow\\"><div class=\\"ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix\\" style=\\"opacity: 1;\\"><div class=\\"ant-select-selection-search\\" style=\\"width: 0px;\\"><input type=\\"search\\" autocomplete=\\"off\\" class=\\"ant-select-selection-search-input\\" role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-owns=\\"rc_select_TEST_OR_SSR_list\\" aria-autocomplete=\\"list\\" aria-controls=\\"rc_select_TEST_OR_SSR_list\\" aria-activedescendant=\\"rc_select_TEST_OR_SSR_list_0\\" readonly=\\"\\" unselectable=\\"on\\" style=\\"opacity: 0;\\" value=\\"\\" id=\\"rc_select_TEST_OR_SSR\\"><span class=\\"ant-select-selection-search-mirror\\" aria-hidden=\\"true\\">&nbsp;</span></div></div></div><span class=\\"ant-select-selection-placeholder\\">Select Agent from the list</span></div></div></div>"`;

exports[`components/form/select/AgentsSelect should render AgentsSelect component with mode=single 1`] = `"<div><div class=\\"ant-select ant-select-single ant-select-allow-clear ant-select-show-arrow ant-select-show-search\\" style=\\"width: 100%;\\"><div class=\\"ant-select-selector\\"><span class=\\"ant-select-selection-search\\"><input type=\\"search\\" autocomplete=\\"off\\" class=\\"ant-select-selection-search-input\\" role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-owns=\\"rc_select_TEST_OR_SSR_list\\" aria-autocomplete=\\"list\\" aria-controls=\\"rc_select_TEST_OR_SSR_list\\" aria-activedescendant=\\"rc_select_TEST_OR_SSR_list_0\\" value=\\"\\" id=\\"rc_select_TEST_OR_SSR\\"></span><span class=\\"ant-select-selection-item\\" title=\\"\\"></span></div><span class=\\"ant-select-arrow\\" style=\\"user-select: none;\\" unselectable=\\"on\\" aria-hidden=\\"true\\"><span role=\\"img\\" aria-label=\\"down\\" class=\\"anticon anticon-down ant-select-suffix\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z\\"></path></svg></span></span><span class=\\"ant-select-clear\\" style=\\"user-select: none;\\" unselectable=\\"on\\" aria-hidden=\\"true\\"><span role=\\"img\\" aria-label=\\"close-circle\\" class=\\"anticon anticon-close-circle\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"close-circle\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z\\"></path></svg></span></span></div></div>"`;
